### ISSUE TITLE
Dependency on array>=0.4.0.1

### ISSUE THREAD

**Issue by TheThirdOne** [idris-lang/Idris-dev#3780]
Running `cabal install` on my machine gives the following dependency conflict.

```
Resolving dependencies...
cabal: Could not resolve dependencies:
trying: idris-1.0 (user goal)
trying: async-2.1.1.1/installed-a68... (dependency of idris-1.0)
next goal: stm (dependency of async-2.1.1.1/installed-a68...)
rejecting: stm-2.2.0.1/installed-91a... (conflict: stm =>
array==0.4.0.0/installed-59d..., idris => array>=0.4.0.1 && <0.6)
rejecting: stm-2.4.4.1, stm-2.4.4, stm-2.4.3, stm-2.4.2, stm-2.4, stm-2.3,
stm-2.2.0.1, stm-2.2.0.0, stm-2.1.2.2, stm-2.1.2.1, stm-2.1.2.0, stm-2.1.1.2,
stm-2.1.1.0, stm-2.1 (conflict: async => stm==2.2.0.1/installed-91a...)
Backjump limit reached (currently 2000, change with --max-backjumps or try to
run with --reorder-goals).
```

This is likely an issue with having an out of date GHC or a similar problem that would be solvable by wiping and reinstalling GHC and cabal.

Is it really necessary to have the to not include array-0.4.0.0? The difference between 0.4.0.0 and 0.4.0.1 is quite minimal. 

As far as I can tell you only use Array, bounds, inRange, and !, which have been in the array package since the first version. So is there a reason the dependency doesn't allow even lower versions?

Relaxing the dependency to `idris => array>=0.4.0.0 && <0.6` reveals another conflict around template-haskell and containers. However, that is a separate issue and I don't  know enough about the versioning to know if that dependency might be able to be relaxed.

**Comment by jfdm** [comment]
> Relaxing the dependency to idris => array>=0.4.0.0 && <0.6 reveals another conflict around template-haskell and containers. However, that is a separate issue and I don't know enough about the versioning to know if that dependency might be able to be relaxed.

Yeah dependency management can be tricky... IIRC there is no reasoning behind the array versioning constraints, most likely it would be to get it working with LTS Haskell.

If you manage to relax the dependencies in a safe way, then by all means contribute with a PR.

**Comment by TheThirdOne** [comment]
So in trying to work out if I can get a working build with relaxed containers and array dependencies, I found that idris uses lookupEnv in Setup.hs, but that isn't guaranteed to exist given the base dependency constraints.

Support for `lookupEnv` was added in 4.6.0.0, but idris allows 4.*

It is a pretty trivial to include, but just requiring a higher base version would make the issues with array and containers both non-issues as they couldn't be used anyway.

Error message:
```
cabal: Entering directory '.'
[1 of 1] Compiling Main             ( dist/dist-sandbox-3ef6f862/setup/setup.hs, dist/dist-sandbox-3ef6f862/setup/Main.o )

dist/dist-sandbox-3ef6f862/setup/setup.hs:159:28:
    Not in scope: `lookupEnv'

dist/dist-sandbox-3ef6f862/setup/setup.hs:161:30:
    Not in scope: `lookupEnv'
```

**Comment by ahmadsalim** [comment]
@TheThirdOne Thanks for reporting the issue. If you want, you are welcome to provide a PR that fixes the constraint or uses another available function.

**Comment by ahmadsalim** [comment]
Since no one provided a PR and no one is complaining about the issue, I will close it.

**Comment by TheThirdOne** [comment]
Just for some context. When I tried to fix update the requirement on my local machine that initially had the issue I ran into more dependency issues. Ultimately there would have been a bunch of small tweaks to make it work (I did get it to install on that machine).

When I tried to replicate it on a new machine I didn't get the issue, it installed correctly first try. So I figured it wasn't worth trying to make sure I didn't break anything for more normal setups because this seems to be a very uncommon bug to hit.
